### PR TITLE
Fxa state machine

### PR DESCRIPTION
**Pull request by bendk**
Adding an FxA state machine in the Rust code.  This one works similarly to the ones in `android-components` and our Swift wrapper.  The main differences are:

- Internal states directly correspond to `FirefoxAccount` method calls.  This makes everything much more testable.  We can write unit tests against the state machine logic without mocking out the `FirefoxAccount` object.  Even better, we can do a dry-run test against the existing implementations before we switch over.  I also think it makes the internal states more meaningful and understandable.
- Clearer separation of which states/events should be visible to the consumer application and which should be internal to the component.
- We always track the last public state that the component was in, so that we can go back to it if we fail in the middle of many internal state transitions.

TODO is the network retry / auth rechecking logic.  I think we can start testing this against the android-components code without that.

This is on top of #5960, let's merge that first.

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/fxa-client/src/state_machine/types.rs [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+///
+/// For each [FxaState] variant, there is a corresponding variant here.  These are called the
+/// public states since they are visible to consumer applications.
+///
+/// There are also variants for internal states.  This states indicate that the state machine is
+/// making a [crate::internal::FirefoxAccount] call. Internal states are temporary.
+/// [crate::FxaStateMachine] will transition to a new state once the call is complete and
+/// before `process_event` completes.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum State {
```
Comment (bendk) [inline components/fxa-client/src/state_machine/types.rs:82]: I initially had a less flattened version of this:

```
pub enum State {
  Public(FxaState),
  Internal(InternalState),
}

pub enum InternalState {
  GetAuthState,
  BeginOAuthFlow { .. },
  ...
}
```

From the view of the types I liked that better, but it made the logic hard to follow because everything was so verbose.

File: components/fxa-client/src/state_machine/mod.rs [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+impl FxaStateMachine {
+    /// Create an FxaStateMachine
+    ///
+    /// Note: When restoring a connected account, only `device_config.capabilities` will be used.
+    /// We will use the device type and device name info that's stored on the server.
+    pub fn new(
+        account: Arc<FirefoxAccount>,
+        oauth_entrypoint: String,
+        device_config: DeviceConfig,
```
Comment (bendk) [inline components/fxa-client/src/state_machine/mod.rs:43]: Maybe we should input the list of OAuth scopes here as well rather than store them in the events.  I didn't do that in case we want to support clients that want flexible permission handling -- for example, they request one set of scopes initially, then upgrade to a new set of scopes later on.

**Review by tarikeshaq — CHANGES_REQUESTED** [review]
Thanks a ton @bendk - I love that we can pull down the complexity from the state machine inwards 🙏 

My main two thoughts:
- What is the high level API a consumer has to use, i.e the functions and types, etc
- The word `State` was already overloaded in the code, and now even more so.. any suggestions?

File: components/fxa-client/src/fxa_client.udl [hunk 3]
```diff
@@ -679,6 +679,24 @@ interface FirefoxAccount {
   void simulate_permanent_auth_token_issue();
 };
 
+interface FxaStateMachine {
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/src/fxa_client.udl:682]: I'm not sure what the best way forward is, but I lean slightly towards let's try to split up `FirefoxAccount` into smaller classes or something like that.  The main reason I say that is that's how the consumers are currently structured.  You normally have one module that uses a state-machine class and another module that uses `DeviceConstellation`.  That said, I think it would also be totally fine to merge it all into one class though too.

File: components/fxa-client/src/state_machine/mod.rs [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+                                },
+                                _ => Event::CallError {
+                                    kind: CallErrorKind::Other,
+                                },
+                            },
+                        },
+                    };
+                    state = next_state(internal_state, event, &current_state)?;
+                }
+            };
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/src/state_machine/mod.rs:101]: A lot of of the remaining comments are about the way the `State` enum is split up.  I think I can do a better job with this, let me try a second draft of these types.

File: components/fxa-client/src/fxa_client.udl [hunk 5]
```diff
@@ -679,6 +679,24 @@ interface FirefoxAccount {
...
+
+  // Get the current auth state
+  FxaState state();
+
+  // Process a auth event (login, logout, etc).
+  //
+  // On success, returns the new state.
+  // On error, the state will remain the same.
+  [Throws=FxaError]
+  FxaState process_event(FxaEvent event);
```
[the remark this answers is not in the document]
Comment (tarikeshaq) [reply components/fxa-client/src/fxa_client.udl:697]: What do you think about _completely_ abstracting the state machine away? i.e the state machine becomes an implementation detail, the consumer can still read the "state", because they have to for presenting UI to the user, etc but from the consumer's perspective, they are triggering procedures (or passing in events in the `process_event` case) but aren't aware of the fact there exists a state machine. Of course, you've spent a considerable time looking at the consumer code so I'd defer to your perspective there.

File: components/fxa-client/src/fxa_client.udl (hunk 3 above)
[the remark this answers is not in the document]
Comment (tarikeshaq) [reply components/fxa-client/src/fxa_client.udl:682]: Yeah I'm trying hard to put my "consumer" hat on, and think about what the best API to interact with is, i.e given what we know now about Android and iOS should the API be a single surface or multiple each for a specific purpose.

(I'm also totally fine exposing this from Rust, I'm mostly thinking about what Fenix/Android Components will consume and we can tie things together in a wapper if we need to)

**Comment by bendk** [comment]
Just pushed a new version of the state machines and I think this one is much cleaner and also documented: https://github.com/mozilla/application-services/blob/ee909b436913a2d028190e087a121cfa0a74c36e/components/fxa-client/src/state_machine/README.md

File: components/fxa-client/src/fxa_client.udl (hunk 3 above)
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/src/fxa_client.udl]: Talked with Tarik about this one and he convinced me that it's better to have a single interface.  Consumer apps may want to split their code into a `device-manager`, `auth-manager`, etc. but that doesn't mean we should do the same.  It's going to be annoying for consumers if we decide on a different split then they want.

TLDR: `process_event` now lives inside `FirefoxAccount` rather than a separate type.

File: components/fxa-client/src/fxa_client.udl (hunk 5 above)
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/src/fxa_client.udl]: This seems right to me.  My hope is that once we're done reworking the android/ios code, that consumers basically just call `process_event()`, see what new state gets returned, and can ignore all the details that happened inside the state machine.

File: components/fxa-client/src/state_machine/internal_machines/mod.rs [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+#[cfg(test)]
+struct StateMachineTester {
+    state_machine: Box<dyn InternalStateMachine>,
+    state: State,
+}
+
+#[cfg(test)]
+impl StateMachineTester {
+    fn new<T: InternalStateMachine + 'static>(state_machine: T, event: FxaEvent) -> Self {
```
Comment (tarikeshaq) [inline components/fxa-client/src/state_machine/internal_machines/mod.rs:185]: nit: If we're using a generic here, can we make the `StateMachineTester` generic as well and throw away the box?

Or alternatively just make this function take the trait object directly

File: components/fxa-client/src/state_machine/internal_machines/mod.rs [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+        device_config: &DeviceConfig,
+    ) -> Result<Event> {
+        let is_ensure_capabilities = matches!(self, State::EnsureDeviceCapabilities);
+        self.make_call_inner(account, device_config).or_else(|e| {
+            // All errors get converted to events, except StateMachineLogicError
+            if matches!(e, Error::StateMachineLogicError(_)) {
+                Err(e)
+            } else {
+                // This call is mostly to report the error, but converting `Error` to `FxaError`
+                // also simplifies some things.
```
Comment (tarikeshaq) [inline components/fxa-client/src/state_machine/internal_machines/mod.rs:106]: nit: can we expand on `some things`?

File: components/fxa-client/src/state_machine/mod.rs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+                event,
+            ),
+            FxaState::Connected => self.process_event_with_internal_state_machine(
+                &internal_machines::ConnectedStateMachine,
+                event,
+            ),
+            FxaState::AuthIssues => self.process_event_with_internal_state_machine(
+                &internal_machines::AuthIssuesStateMachine,
+                event,
+            ),
```
Comment (tarikeshaq) [inline components/fxa-client/src/state_machine/mod.rs:55]: nit: can we use the match to get our state machine, then call `process_event_with_internal_state_machine` after? (so the call is only there once and not duplicated per match arm)

File: components/fxa-client/src/state_machine/mod.rs [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+        }
+    }
+
+    fn process_event_with_internal_state_machine<T: InternalStateMachine>(
+        &mut self,
+        state_machine: &T,
```
Comment (tarikeshaq) [inline components/fxa-client/src/state_machine/mod.rs:61]: nit: does this need to be a reference? the state_machines are short lived, so I would have expected us to pass ownership to this function

**Review by tarikeshaq — APPROVED** [review]
Thank you so much for driving this @bendk , I love how this is looking!! 

I had a few comments but mostly about followups and nits. Please feel free to re-request my review if you make substantial changes, but otherwise this looks good! I'm curious to see how this hooks up into the app

File: components/fxa-client/src/device.rs [hunk 10]
```diff
@@ -197,7 +197,7 @@ impl FirefoxAccount {
 }
 
 /// Device configuration
-#[derive(Debug)]
+#[derive(Clone, Debug, PartialEq, Eq)]
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/src/device.rs:200]: It's needed because `DeviceConfig` is a field in the `Event::Initialize` and those need to be compared (maybe only for the tests though, I'm not sure).

File: components/fxa-client/src/internal/mod.rs [hunk 11]
```diff
@@ -54,6 +54,11 @@ pub struct FirefoxAccount {
     devices_cache: Option<CachedResponse<Vec<http_client::GetDeviceResponse>>>,
     auth_circuit_breaker: AuthCircuitBreaker,
     telemetry: FxaTelemetry,
+    // TODO: Cleanup our usage of the word "state" and change this field name to `state`
+    // https://bugzilla.mozilla.org/show_bug.cgi?id=1868610
+    pub(crate) fxa_state: FxaState,
+    // Set via `FxaEvent::Initialize`
+    pub(crate) device_config: Option<DeviceConfig>,
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/src/internal/mod.rs:61]: Yeah, I think this would be great.  https://bugzilla.mozilla.org/show_bug.cgi?id=1868895

File: components/fxa-client/src/state_machine/README.md [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+
+Here is an overview containing some of the states and transitions:
+
+```mermaid
+graph LR;
+    Disconnected --> |"BeginOAuthFlow(Success)"| Authenticating
+    Disconnected --> |"BeginOAuthFlow(Failure)"| Disconnected
+    Disconnected --> |"BeginPairingFlow(Success)"| Authenticating
+    Disconnected --> |"BeginPairingFlow(Failure)"| Disconnected
+    Authenticating --> |"CompleteOAuthFlow(Success)"| Athenticated
```
[the remark this answers is not in the document]
Comment (bendk) [reply components/fxa-client/src/state_machine/README.md:20]: Actually `Connected`, my typo broke find/replace :)

File: components/fxa-client/src/state_machine/mod.rs (hunk 9 above)
Comment (bendk) [reply components/fxa-client/src/state_machine/mod.rs:61]: I guess I was thinking of them as long-lived, but zero-sized, types.  If for some reason we wanted to store state in the state machines themselves, then we'd probably want to input a reference rather than an owned value.  That said, we're not doing that so let's remove the `&` symbol.

File: components/fxa-client/src/state_machine/mod.rs (hunk 8 above)
Comment (tarikeshaq) [reply components/fxa-client/src/state_machine/mod.rs:55]: Blah that's right, don't know why I thought we were making trait objects 🙃
